servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java

```diff
@@ -374,6 +396,38 @@ default HttpServerContext listenBlockingStreamingAndAwait(BlockingStreamingHttpS
         return blockingInvocation(listenBlockingStreaming(service));
     }
 
+    /**
+     * Starts this server and returns the {@link HttpServerContext} after the server has been successfully started.
+     * <p>
+     * If the underlying protocol (e.g. TCP) supports it this will result in a socket bind/listen on {@code address}.
+     * <p>
+     * Note that this method is generic in the sense that it accepts all HTTP {@link Service} implementations to be
+     * passed in, namely {@link StreamingHttpService}, {@link HttpService}, {@link BlockingStreamingHttpService} and
+     * {@link BlockingHttpService}. It is especially useful when Dependency Injection is used and the type of service
+     * is not known at compile time.
+     *
+     * @param service Service invoked for every request received by this server. The returned {@link HttpServerContext}
+     * manages the lifecycle of the {@code service}, ensuring it is closed when the {@link HttpServerContext} is closed.
+     * @return A {@link Single} that completes when the server is successfully started or terminates with an error if
+     * the server could not be started.
+     * @throws IllegalArgumentException if an unknown/unsupported {@link Service} is being provided.
+     */
+    default Single<HttpServerContext> listenService(final Service service) {
+        Objects.requireNonNull(service);
+
+        if (service instanceof HttpService) {
+            return listen((HttpService) service);
+        } else if (service instanceof StreamingHttpService) {
+            return listenStreaming((StreamingHttpService) service);
+        } else if (service instanceof BlockingHttpService) {
+            return listenBlocking((BlockingHttpService) service);
+        } else if (service instanceof BlockingStreamingHttpService) {
+            return listenBlockingStreaming((BlockingStreamingHttpService) service);
+        } else {
+            throw new IllegalArgumentException("Unknown/Unsupported service type: " + service);
```

Comment:
1. Methods that return an async source should not throw (unless NPE). Instead, those methods should returns a failed source. See `Single.failed(...)`.
2. Consider keeping only one variant, let's say "Unsupported".